trin-core/src/portalnet/overlay.rs

```diff
@@ -152,20 +142,227 @@ impl Overlay {
-    pub fn table_entries_enr(&self) -> Vec<Enr> {
+    pub async fn table_entries_enr(&self) -> Vec<Enr> {
         self.kbuckets
             .write()
+            .await
             .iter()
             .map(|entry| entry.node.value.enr().clone())
             .collect()
     }
+
+    pub async fn send_ping(&self, data_radius: U256, enr: Enr) -> Result<Vec<u8>, String> {
+        let enr_seq = self.discovery.read().await.local_enr().seq();
+        let msg = Ping {
+            enr_seq,
+            data_radius,
+        };
+        self.discovery
+            .write()
+            .await
+            .send_talkreq(
+                enr,
+                PROTOCOL.to_string(),
+                Message::Request(Request::Ping(msg)).to_bytes(),
+            )
+            .await
+    }
+
+    pub async fn send_find_nodes(&self, distances: Vec<u16>, enr: Enr) -> Result<Vec<u8>, String> {
+        let msg = FindNodes { distances };
+        self.discovery
+            .write()
+            .await
+            .send_talkreq(
+                enr,
+                PROTOCOL.to_string(),
+                Message::Request(Request::FindNodes(msg)).to_bytes(),
+            )
+            .await
+    }
+
+    pub async fn send_find_content(
+        &self,
+        content_key: Vec<u8>,
+        enr: Enr,
+    ) -> Result<Vec<u8>, String> {
+        let msg = FindContent { content_key };
+        self.discovery
+            .write()
+            .await
+            .send_talkreq(
+                enr,
+                PROTOCOL.to_string(),
+                Message::Request(Request::FindContent(msg)).to_bytes(),
+            )
+            .await
+    }
+}
+
+#[derive(Clone)]
+pub struct HistoryProtocol {
+    pub overlay: Arc<OverlayProtocol>,
+}
+
+#[derive(Clone)]
+pub struct StateProtocol {
+    pub overlay: Arc<OverlayProtocol>,
+}
+
+impl StateProtocol {
+    pub async fn new(
+        discovery: Arc<RwLock<Discovery>>,
+        db: Arc<DB>,
+        portal_config: PortalnetConfig,
+    ) -> Result<(Self, PortalnetEvents), String> {
+        let config = OverlayConfig::default();
+        let kbuckets = Arc::new(RwLock::new(KBucketsTable::new(
+            discovery.read().await.local_enr().node_id().into(),
+            config.bucket_pending_timeout,
+            config.max_incoming_per_bucket,
+            config.table_filter,
+            config.bucket_filter,
+        )));
+        let data_radius = Arc::new(RwLock::new(portal_config.data_radius));
+
+        let protocol_receiver = discovery
+            .write()
+            .await
+            .discv5
+            .event_stream()
+            .await
+            .map_err(|e| e.to_string())
+            .unwrap();
+
+        let overlay = OverlayProtocol {
+            discovery: discovery.clone(),
+            data_radius,
+            kbuckets,
+        };
+
+        let overlay = Arc::new(overlay);
+
+        let utp_listener = UtpListener {
+            discovery: discovery.clone(),
```

Comment:
With this setup, we're cloning a lot of arcs - which seems fine, but I'm just curious how you feel about being more explicit in our arc variable naming? ie. `discovery_arc` / `overlay_arc`. I don't love it, but it's hard to keep track of all the arc's floating about, and we want to avoid accidentally cloning the base datatype. This isn't really a suggestion, just thinking out loud. It's reasonable that the compiler helps us out with this so we don't have to worry about being so explicit

Reply:
I'm not sure what are the Rust's best practices here..., but I think the type checking does a great job and it will not allow us to pass base datatype if it is not expected...